src/apps/replication/zookeeper/distributed_lock_service_zookeeper.cpp

```diff
@@ -0,0 +1,231 @@
+        else
+            handle = iter->second;
+    }
+
+    auto lock_tsk = tasking::create_late_task<distributed_lock_service::lock_callback>(
+        lock_cb_code, 
+        lock_cb
+    );
+    auto expire_tsk = tasking::create_late_task<distributed_lock_service::lock_callback>(
+        lease_expire_code,
+        lease_expire_callback
+    );
+    
+    tasking::enqueue(TASK_CODE_DLOCK, nullptr, std::bind(&lock_struct::try_lock, handle, lock_tsk, expire_tsk), handle->hash());
+    return std::make_pair(lock_tsk, expire_tsk);
+}
+
+task_ptr distributed_lock_service_zookeeper::unlock(
+    const std::string& lock_id, 
+    const std::string& myself_id, 
+    bool destroy, 
+    task_code cb_code, 
+    const err_callback& cb)
+{
+    lock_struct_ptr handle;
+    {
+        utils::auto_read_lock l(_service_lock);
+        auto iter = _zookeeper_locks.find( std::make_pair(lock_id, myself_id) );
+        if (iter == _zookeeper_locks.end())
+            return tasking::enqueue(cb_code, nullptr, std::bind(cb, ERR_OBJECT_NOT_FOUND));
+        handle = iter->second;
+    }
+    auto unlock_tsk = tasking::create_late_task<distributed_lock_service::err_callback>(cb_code, cb);
+    tasking::enqueue(TASK_CODE_DLOCK, nullptr, std::bind(&lock_struct::unlock, handle, unlock_tsk), handle->hash());
+    return unlock_tsk;
+}
+
+task_ptr distributed_lock_service_zookeeper::cancel_pending_lock(
+    const std::string& lock_id,
+    const std::string& myself_id,
+    task_code cb_code,
+    const lock_callback& cb)    
+{
+    lock_struct_ptr handle;
+    {
+        utils::auto_read_lock l(_service_lock);
+        auto iter = _zookeeper_locks.find( std::make_pair(lock_id, myself_id) );
+        if (iter == _zookeeper_locks.end())
+            return tasking::enqueue(cb_code, nullptr, std::bind(cb, ERR_OBJECT_NOT_FOUND, "", -1));
+        handle = iter->second;        
+    }
+    auto cancel_tsk = tasking::create_late_task<distributed_lock_service::lock_callback>(cb_code, cb);
+    tasking::enqueue(TASK_CODE_DLOCK, nullptr, std::bind(&lock_struct::cancel_pending_lock, handle, cancel_tsk), handle->hash());
+    return cancel_tsk;
+}
+
+task_ptr distributed_lock_service_zookeeper::query_lock(
+    const std::string& lock_id, 
+    task_code cb_code, 
+    const lock_callback& cb)
+{
+    lock_struct_ptr handle = nullptr;
+    {
+        utils::auto_read_lock l(_service_lock);
+        for (auto& lock_pair: _zookeeper_locks)
+        {
+            if (lock_pair.first.first == lock_id)
+            {
+                handle = lock_pair.second;
+                break;
+            }
+        }
+    }
+    if (handle == nullptr)
+        return tasking::enqueue(cb_code, nullptr, std::bind(cb, ERR_OBJECT_NOT_FOUND, "", -1)); 
+    else {
+        auto query_tsk = tasking::create_late_task<distributed_lock_service::lock_callback>(cb_code, cb);
+        tasking::enqueue(TASK_CODE_DLOCK, nullptr, std::bind(&lock_struct::query, handle, query_tsk), handle->hash() );
```

Comment:
better to use separated task code for different callbacks - otherwise all underlying tools (tracing/profiling/web-monitor/explore/fault-injector/ etc.) cannot differentiate them.